external/kafka/src/main/scala/io/gearpump/streaming/kafka/KafkaSource.scala

```diff
@@ -161,18 +161,20 @@ class KafkaSource(
     setStartTime(startTime)
   }
 
-  override def read(batchSize: Int): List[Message] = {
-    val messageBuffer = ArrayBuffer.empty[Message]
-
-    fetchThread.foreach {
-      fetch =>
-        var count = 0
-        while (count < batchSize) {
-          fetch.poll.flatMap(filterMessage).foreach(messageBuffer += _)
+  override def read(batchSize: Int): Array[Message] = {
+    var count = 0
+    val messages = new Array[Message](batchSize)
+    fetchThread.foreach { fetch =>
+      (0 until batchSize).foreach { _ =>
+        fetch.poll.flatMap(filterMessage).foreach { message =>
+          messages(count) = message
           count += 1
         }
+      }
     }
-    messageBuffer.toList
+    if (count < batchSize) {
+      messages.take(count)
+    } else messages
```

Comment:
I'm a bit worried about this. 

What if 

* a user takes an array but reads an index that doesn't exist ?
* a user `read` twice ? It seems the behavior is undetermined. It could be the original array. It could be a shrunk copy.